application-analytics-default/src/main/java/com/xwiki/analytics/internal/MostViewedJsonNormaliser.java

```diff
@@ -0,0 +1,207 @@
+
+/**
+ * Implementation for {@link JsonNormaliser}.
+ *
+ * @version $Id$
+ * @since 1.0
+ */
+@Component
+@Named("MostViewedPages")
+@Singleton
+public class MostViewedJsonNormaliser implements JsonNormaliser
+{
+    /**
+     * Hint for the MostViewedJsonNormaliser.
+     */
+    public static final String HINT = "MostViewedPages";
+    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+
+    private static final String DATE = "date";
+
+    private static final String URL = "url";
+
+    private static final String LABEL = "label";
+
+    @Inject
+    private ResourceReferenceResolver<ExtendedURL> resourceReferenceResolver;
+
+    @Inject
+    private Logger logger;
+
+    @Inject
+    private ResourceTypeResolver<ExtendedURL> resourceTypeResolver;
+
+    @Inject
+    @Named("compactwiki")
+    private EntityReferenceSerializer<String> serializer;
+
+    /**
+     * Normalize Matomo response for format consistency and add extra information needed by XWiki.
+     *
+     * @param jsonString the Matomo JSON response, in {@code String} format
+     * @return the normalised json
+     */
+    public JsonNode normaliseData(String jsonString) throws JsonProcessingException
+    {
+        // Convert the string returned by Matomo in a JSON format to easily handle the processing of the nodes.
+        JsonNode jsonRoot = OBJECT_MAPPER.readTree(jsonString);
+        // Matomo may return several variants of JSON formats. In one scenario, when the period is set to
+        // day/week/month/year, it returns a JSON object with keys representing dates. The corresponding value for
+        // each key is an array of JSON objects, each of which represents a page. However, if the user sets the
+        // period parameter to "range" Matomo returns an array of JSON objects, with each JSON object representing
+        // a page. Due to these variations, the result needs to be processed to create a single format.
+        // This normalized format is an array of JSON objects and each JSON object in this array will have a new
+        // field called 'date'. This 'date' field will be set to 'N/A' when Matomo returns an array instead of an
+        // object. For both type of formats, the label field is also altered in order to contain the full page name
+        if (jsonRoot.isArray()) {
+            processArrayNode(jsonRoot);
+        } else {
+            jsonRoot = processObjectNode(jsonRoot);
+        }
+        return jsonRoot;
+    }
+
+    /**
+     * This method processes each entry to append an empty date to it.
+     *
+     * @param jsonNode an array of jsons
+     */
+    private void processArrayNode(JsonNode jsonNode)
+    {
+        for (JsonNode objNode : jsonNode) {
+            if (objNode.isObject()) {
+                ((ObjectNode) objNode).put(DATE, "");
+                // Just in case that this normalsier wil be used in the future for other macros that do not have an url.
```

Comment:
That was just a comment I had, to keep in mind for future changes. No need to add it here, it does not help right now. (to be removed below as well)
What I said to keep in mind is that, again for future uses of this normalizer, that there might be other nodes that have a `url`, besides the label node, so maybe `if (objNode.has(URL))` will not point to the right node. But this is not the case right now